### PR TITLE
Add Placing Parentheses and update README

### PR DESCRIPTION
**Pull request by srivastavaayu**
Added Placing Parenthesis (the question is to maximize an expression by rearranging operands and operators) in the Dynamic Programming subdirectory.
Add repo license details with a link to the license file in the directory.

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [x] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: dynamic_programming/parentheses.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    M=[[None for x in range(n)] for x in range(n)]
+    for i in range(n):
+        m[i][i]=operand[i]
+        M[i][i]=operand[i]
+    for s in range(1,n):
+        for i in range(0,n-s):
+            j=i+s
+            m[i][j],M[i][j]=MinandMax(M,m,i,j,operator)
+    return M[0][n-1]
+expression=input()
```
Comment (cclauss) [inline dynamic_programming/parentheses.py:36]: This line will hang our automated testing as discussed in CONTRIBUTING.md.

**Comment by cclauss** [comment]
Codespell says:
```
 ./dynamic_programming/parentheses.py:24: parantheses ==> parentheses
./dynamic_programming/parentheses.py:43: parantheses ==> parentheses
```

**Comment by cclauss** [comment]
https://docs.python.org/3/library/operator.html
```python
from operator import add, subtract, multiply, true_division

operators = {
    "+": add,
    "-": subtract,
    "*": multiply,
    "/": floor_division  # true_division,
}


def calculate(a: int, b: int, op: str) -> int:
    return operators[op](a, b)
```

File: dynamic_programming/parentheses.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    """
+    if op == "+":
+        return a + b
+    elif op == "-":
+        return a - b
+    else:
+        return a * b
+
+
+def MinandMax(M, m, i, j, operator) -> int:
```
Comment (cclauss) [inline dynamic_programming/parentheses.py:16]: Single letter variable names are old school.  Python naming conventions say that capital letters are for constants.

File: parentheses.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import math
+from operator import add, sub, mul
+
+
+operators={"+": add, "-": sub, "*": mul}
+
+
+def calculate(first_operand, second_operand, operator) -> int:
```
Comment (cclauss) [inline parentheses.py:8]:
```suggestion
def calculate(first_operand: int, second_operand: int, operator: str) -> int:

```

File: parentheses.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+
+def calculate(first_operand, second_operand, operator) -> int:
+    """
+    Return the value after evaluating the expression
+    """
+    return operators[operator](first_operand, second_operand)
+
+
+def MinandMax(maximum_numbers, minimum_numbers, i, j, operator) -> int:
```
Comment (cclauss) [inline parentheses.py]:
```suggestion
def min_and_max(minimum_numbers: float, maximum_numbers: float, i: int, j: int, operator: str) -> int:

```

Comment (cclauss) [reply parentheses.py]: If the function is called `min_and_max()` then the order of the parameters can not be the opposite.

File: parentheses.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        for i in range(0, number_of_operands - s):
+            j = i + s
+            minimum_numbers[i][j], maximum_numbers[i][j] = MinandMax(maximum_numbers, minimum_numbers, i, j, operator)
+    return maximum_numbers[0][number_of_operands - 1]
+
+
+expression = input().strip()
+operator, operand = [], []
+for i in expression:
+    if i in operators.keys():
```
Comment (cclauss) [inline parentheses.py]:
```suggestion
    if i in operators:

```

File: parentheses.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        operator.append(i)
+    else:
+        operand.append(int(i))
+answer = parentheses(operand, operator)
+f"{answer}"
```
Comment (cclauss) [inline parentheses.py]:
```suggestion
f"{parentheses(operand, operator)}"
```

File: parentheses.py (hunk 5 above)
Comment (cclauss) [inline parentheses.py]:
```suggestion
    if i in operators:

```

**Comment by srivastavaayu** [comment]
@cclauss I don't understand. I've possibly done every change suggested by you. Let me try and incorporate changes suggested by you once again. But is there any specific reason the build is failing?

File: parentheses.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        minimum_numbers[i][i] = operand[i]
+        maximum_numbers[i][i] = operand[i]
+    for s in range(1, number_of_operands):
+        for i in range(0, number_of_operands - s):
+            j = i + s
+            minimum_numbers[i][j], maximum_numbers[i][j] = MinandMax(maximum_numbers, minimum_numbers, i, j, operator)
+    return maximum_numbers[0][number_of_operands - 1]
+
+
+expression = input().strip()
```
[the remark this answers is not in the document]
Comment (srivastavaayu) [reply parentheses.py]: @cclauss I don't understand what you want me to do here. I've gone through CONTRIBUTING.md but I've done almost all the necessary changes I could. Could you please elaborate the issue?

File: dynamic_programming/parentheses.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    number_of_operands = len(operand)
+    minimum_numbers = [
+        [None for x in range(number_of_operands)] for x in range(number_of_operands)
+    ]
+    maximum_numbers = [
+        [None for x in range(number_of_operands)] for x in range(number_of_operands)
+    ]
+    for i in range(number_of_operands):
+        minimum_numbers[i][i] = operand[i]
+        maximum_numbers[i][i] = operand[i]
```
Comment (cclauss) [inline dynamic_programming/parentheses.py:41]:
```suggestion
    number_of_operands = len(operands)
    minimum_numbers = [[None] * number_of_operands] * number_of_operands
    maximum_numbers = [[None] * number_of_operands] * number_of_operands
    for i, operand in enumerate(operands):
        minimum_numbers[i][i] = operand
        maximum_numbers[i][i] = operand
```

File: dynamic_programming/parentheses.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+expression = input().strip()
+operator, operand = [], []
+for i in expression:
+    if i in operators.keys():
+        operator.append(i)
+    else:
+        operand.append(int(i))
+answer = parentheses(operand, operator)
+f"{answer}"
```
Comment (cclauss) [inline dynamic_programming/parentheses.py:59]:
```suggestion
if __name__ == "__main__":
    expression = input("Please provide a mathimatical expression... ").strip()
    operator, operand = [], []
    for c in expression:
        if c in operator:
            operator.append(c)
        else:
            operand.append(int(c))

    print(f"The result is {parentheses(operand, operator)}")
```

Comment (cclauss) [reply dynamic_programming/parentheses.py:59]: Use `c` for characters and `i`, `j` for integers.

**Comment by srivastavaayu** [comment]
@cclauss All the tests have passed successfully. Please review this PR. And suggest changes, if any.

File: dynamic_programming/parentheses.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+            minimum_numbers[i][j], maximum_numbers[i][j] = MinandMax(
+                maximum_numbers, minimum_numbers, i, j, operator
+            )
+    return maximum_numbers[0][number_of_operands - 1]
+
+
+if __name__ == "__main__":
+    expression = input("Please provide a mathematical expression... ").strip()
+    operator, operands = [], []
+    for c in expression:
```
Comment (cclauss) [inline dynamic_programming/parentheses.py:61]:
```suggestion
    for c in expression.replace(" ", ""):
```
Deal with spaces in the expression.

File: dynamic_programming/parentheses.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    4
+    >>> calculate(2,2,"-")  # sub()
+    0
+    >>> calculate(2,2,"*")  # mul()
+    4
+    """
+    return operators[operator](first_operand, second_operand)
+
+
+def MinandMax(maximum_numbers, minimum_numbers, i, j, operator) -> int:
```
Comment (cclauss) [inline dynamic_programming/parentheses.py:22]:
```suggestion
def MinandMax(maximum_numbers: List[int], minimum_numbers: List[int], i: int, j: int, operator: str) -> int:
```
Please do not mark this resolved until the function definition looks like this and has doctexts.

File: dynamic_programming/parentheses.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
+import math
+from operator import add, sub, mul
+
```
Comment (cclauss) [inline dynamic_programming/parentheses.py:3]:
```suggestion
from typing import List

```

**Comment by cclauss** [comment]
This needs a doctest that proves that this code is ___rearranging operands and operators___.  It also needs a URL to a source that discusses the dynamic programming aspects of this algorithm.

**Pull request closed without merging**